### ISSUE TITLE
Receive events on shape create and delete

### ISSUE THREAD

**Issue by cudmore** [napari/napari#720]
## 🚀 Feature
Can we get events when shape layer shapes are created and deleted?

## Motivation

I am implementing a shape analysis plugin. It would be great if my code could receive an event when a shape is added and a shape is deleted.

I want to display the current shapes in list, something like a Qt TreeView. I need to append and delete from this list as shapes are added and deleted from the Napari interface. To do this, I need to receive events when a new shape is created and a shape is deleted.

## Pitch

Because there can be many shapes (e.g. ROIs), it is super useful to display them as a list. In a Qt TreeView for example. With this, the user can see the number of shapes, the properties of each (including my analysis results), can select them in the list and have them selected in the viewer, can delete them, etc. etc.

When there are many shapes, it is not useful to only display them on top of an image. This results in **user fatigue**. It is useful to have them in a list so the user can browse shapes from the list as compared to a visual search on top of the image. Thus, I need Napari to trigger events on shape addition and deletion.

When shapes/ROIs are only on top of the image, it becomes similar to [where is Waldo](https://www.google.com/search?q=where+is+waldo&sxsrf=ACYBGNSnqjLq7LwPJcsMs8nydf4f6HKJ2Q:1574393515050&source=lnms&tbm=isch&sa=X&ved=2ahUKEwi9xs3x8PzlAhVIGDQIHYWWCBoQ_AUoAXoECA4QAw&biw=1258&bih=1231#imgrc=0W6qdOm2E9qAdM:)!

## Alternatives

I can find shape and shape layer events triggered when properties of a shape change, things like (edge_color, face_color, edge_width). But no code that generates an event when a shape is created or deleted?

 - New shape

For new shapes I am catching mouse down events using the `mouse_drag_callbacks` decorator. But this requires me to keep a backend list of current shapes and compare to the list when the mouse is clicked.

```
@self.shapeLayer.mouse_drag_callbacks.append
def shape_mouse_move_callback(layer, event):
    if event.type == 'mouse_press':
        #compare list of shapes in my backend shape layer
        #if there are more than in my backend then the last one is a new one
        pass
```

 - Delete shape

I am not sure how to do this? I found `napari.layers.shapes.shapeList.remove()` which does not trigger an event and seems to be called repeatedly as shape properties are updated? Properties like (edge_color, face_color, edge_width). Maybe it should be `_remove()`. I also found `remove_selected()` but it also does not trigger an event?

## Additional context

I am storing the analysis for each shape in the shape `metadata` data member. With this, my code does not need to know about added or deleted shapes. But I would like to keep a backend list of shapes to be displayed in a Tree View so the user can select shapes from the list as compared to selecting in the image.

![cudmore-napari-plugin-1](https://user-images.githubusercontent.com/1009168/69394311-825a4780-0c90-11ea-91f9-9b839b1c2f3c.gif)

On the left is the Napari viewer with two shapes, a line in blue and a polygon/rectangle in orange. On the right is my PyQtGraph interface for shape analysis. It has 4 plots., (i) a line intensity profile (white) along with my analysis including a gaussian fit (red) and my diameter estimate (blue). This is all updated in real time as the user drags a selected line shape and/or adjusts the viewed image plane/frame, (ii) my meta analysis of the diameter along the line, (iii) a plot of all the line profiles through the time series, and (iv) a plot of the mean intensity within the polygon/rectangle napari shape.

I would like to add a list of current shapes/ROIs to this interface. Something like a Qt TreeView. To do this, I need events that tell me when a shape is added and a shape is deleted...

**Comment by sofroniewn** [comment]
Hi @cudmore thanks for the very through issue on this. I've been meaning to add better events to the shape layer for a while, for example see #517 but havn't done that yet.

I think we should come up now with a list of all the events we want to add to the shape layer and standardize what information we want that event to carry.

For example we should probably emit events when
- [ ] new shape is added
- [ ] shape is deleted

but do we also want events when
- [ ] shape drawing starts
- [ ] shape drawing ends
- [ ] vertex is added / removed
- [ ] shape is moved
etc.

and what information do we want to emit with the event. We could have `event.shape` carry the `Shape`, or `event.index` carry the index of the shape added / deleted.

Maybe we start simple with just `new_shape` and `delete_shape`, but I want to raise these questions now before we implement anything

**Comment by sofroniewn** [comment]
@cudmore also if you are interested in contributing this addition we can help you through the process, could be a great first PR to napari for you!!

**Comment by Asharits** [comment]
> This issue has been mentioned on **Image.sc Forum**. There might be relevant details there:
> 
> https://forum.image.sc/t/event-emitter-for-deleting-data-within-a-layer-points-shapes-etc/51853/3

@sofroniewn, so to continue the conversation here. The issue I have is that I am creating a shapes layer for each object I want to label. This makes some of the data processing I want to do later easier. @jni, correct me if I am wrong but I think both your suggestions only work if you have one shapes layer to watch. I found with the way I have written my code, I need to know which layer is selected first before I can capture where data has been added. I am sure there is probably a better way to write my code so that i can utilize the existing API. 
```python
import napari
import numpy as np
from pathlib import Path
from magicgui import magicgui
from skimage.data import brain
from napari.layers import Shapes
from magicgui.widgets import TextEdit,LineEdit,Table

def create_annotate_widgets(viewer):

    @magicgui(call_button="create new label",
              orientation_choice={'choices': ('vertical','horizontal','diagonal','N/A')},
              result_widget=True)
    def create_label(new_label_name="", orientation_choice='vertical')-> Shapes:
        if new_label_name in viewer.layers:
            return 'Label already exists!'
        else:
            metadata_dict = {'label name': new_label_name, 'orientation': orientation_choice}
            return Shapes(name=new_label_name,metadata=metadata_dict,ndim = 3,edge_width=2,edge_color=np.random.random(3),face_color='transparent')

    @magicgui(call_button='Save annotation data',
          label_name={'widget_type':'LineEdit','label':'label name','enabled': False}, 
          orientation_label={'widget_type':'LineEdit','label':'orientation','enabled': False},
          data_list={'widget_type':'Table','label':'Data'},
          save_dir={'widget_type':'LineEdit','label':'filename'},
          result_widget=True)
    def annotation_menu(label_name, orientation_label, data_list,save_dir):
        """
        Insert code for saving to dataframe
        """
        return 'Data saved'

    def update_annotation_menu(event):
        annotation_menu.label_name.value = event.value.metadata['label name']
        annotation_menu.orientation_label.value = event.value.metadata['orientation']
        data = event.value.data
        bbox_list = sorted([[box[0][0],box[0][2],box[2][2],box[0][1],box[2][1]] for num,box in enumerate(data)])
        index = [f'box {i}' for i in range(len(data))]
        annotation_menu.data_list.value = {'data':bbox_list,'index':index ,'columns':('Slice','X-min','X-max','Y-min','Y-max')}

    def add_data(viewer, event):
        yield
        while event.type == 'mouse_move':
            yield
        data = viewer.layers.selection.active.data
        bbox_list = sorted([[box[0][0],box[0][2],box[2][2],box[0][1],box[2][1]] for num,box in enumerate(data)])
        index = [f'box {i}' for i in range(len(data))]
        annotation_menu.data_list.value = {'data':bbox_list,'index':index ,'columns':('Slice','X-min','X-max','Y-min','Y-max')}

    viewer.layers.selection.events._current.connect(update_annotation_menu)
    viewer.mouse_drag_callbacks.append(add_data)

    return create_label, annotation_menu

def annotate_data(data):
    viewer = napari.Viewer()
    viewer.add_image(data)
    create_label, annotation_menu = create_annotate_widgets(viewer)
    viewer.window.add_dock_widget(create_label, area='bottom')
    viewer.window.add_dock_widget(annotation_menu, area='right')
    napari.run()

annotate_data(brain())
```

My quick fix is similar to what @cudmore suggested above in keeping a backend list of the data.  The difference is that I am using a dataclass which allows me to capture adding and deleting but as I mentioned in my post I would prefer to utilize your native data structure since the data is already there. The other thought I had was nested callbacks but I have not gone down that road yet.

**Comment by sofroniewn** [comment]
> My quick fix is similar to what @cudmore suggested above in keeping a backend list of the data. The difference is that I am using a dataclass which allows me to capture adding and deleting but as I mentioned in my post I would prefer to utilize your native data structure since the data is already there.

Agreed I think it would be much nicer for you to be able to use our data structure here. @kevinyamauchi might have some good ideas here. This could also be something for @DragaDoncila to work on too

**Comment by DragaDoncila** [comment]
I think emitting when we `add` a shape would be easy since we compute the number of new shapes to add both in the `data` setter and in the `add` method. Same with `remove_selected`, we could emit from there. 

Exposing the `index` of the shape(s) in question with the event seems very useful if users are maintaining associated tables of information about the shapes e.g. analysis results.

**Comment by dstansby** [comment]
Looks like this can be closed by https://github.com/napari/napari/pull/5967 now?

**Comment by psobolewskiPhD** [comment]
Yeah, I think with:
- https://github.com/napari/napari/pull/2992
- https://github.com/napari/napari/pull/5967 
- https://github.com/napari/napari/pull/6178
all of the cases above are covered?
